engine/consensus/sealing/core.go

```diff
@@ -36,1001 +28,401 @@ import (
-		Int64("duration_ms", time.Since(startTime).Milliseconds()).
-		Msg("checking sealing finished successfully")
-
 	return nil
 }
 
-// sealableResults returns the IncorporatedResults from the mempool that have
-// collected enough approvals on a per-chunk basis, as defined by the matchChunk
-// function. It also filters out results that have an incorrect sub-graph.
-// It specifically returns the information for the next unsealed results which will
-// be useful for debugging the potential sealing halt issue
-func (c *Core) sealableResults() (flow.IncorporatedResultList, *tracker.SealingTracker, error) {
-	// tracker to collection information about the _current_ sealing check.
-	sealingTracker := tracker.NewSealingTracker(c.state)
-
-	lastFinalized, err := c.state.Final().Head()
+// processApproval implements business logic for processing single approval
+// Returns:
+// * engine.InvalidInputError - result approval is invalid
+// * engine.UnverifiableInputError - result approval is unverifiable since referenced block cannot be found
+// * engine.OutdatedInputError - result approval is outdated for instance block was already sealed
+// * exception in case of any other error, usually this is not expected
+// * nil - successfully processed result approval
+func (c *Core) processApproval(approval *flow.ResultApproval) error {
+	err := c.checkBlockOutdated(approval.Body.BlockID)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get last finalized block: %w", err)
+		return fmt.Errorf("won't process approval for oudated block (%x): %w", approval.Body.BlockID, err)
 	}
 
-	// go through the results mempool and check which ones we can construct a candidate seal for
-	var results []*flow.IncorporatedResult
-	for _, incorporatedResult := range c.incorporatedResults.All() {
-		// Can we seal following the happy-path protocol, i.e. do we have sufficient approvals?
-		sealingStatus, err := c.hasEnoughApprovals(incorporatedResult)
-		if state.IsNoValidChildBlockError(err) {
-			continue
-		}
-		if err != nil {
-			return nil, nil, fmt.Errorf("internal error sealing chunk approvals to incorporated result: %w", err)
-		}
-		sealableWithEnoughApprovals := sealingStatus.SufficientApprovalsForSealing
-		sealingTracker.Track(sealingStatus)
-
-		// Emergency Sealing Fallback: only kicks in if we can't seal following the happy-path sealing
-		emergencySealable := false
-		if !sealableWithEnoughApprovals {
-			emergencySealable, err = c.emergencySealable(incorporatedResult, lastFinalized)
-			if err != nil {
-				return nil, nil, fmt.Errorf("internal error sealing chunk approvals to incorporated result: %w", err)
-			}
-			sealingStatus.SetQualifiesForEmergencySealing(emergencySealable)
+	if collector, processable := c.collectorTree.GetCollector(approval.Body.ExecutionResultID); collector != nil {
+		if !processable {
+			return engine.NewOutdatedInputErrorf("collector for %s is marked as non processable", approval.Body.ExecutionResultID)
```

Comment:
I think the current code would work, though it behaves a little bit strange in my opinion:
* we cache approvals for blocks, whose result we haven't seen yet
* when the result comes along (but is detached, e.g. because we are missing an ancestor result), we throw away all the approvals because the `AssignmentCollector` is not processable. 

Overall, I feel we should still cache the approval (in the collector), even if the collector is not yet processable, because it might become processable in the future. 

This goes back to the three different states a collector can be in: 
1. not yet processable
2. processable 
3. orphaned 

Previously, we tried to represent these states with a boolean variable. I see different ways forward:
* Option (i): we keep the boolean variable. Then we would need two different caches for each `AssignmentCollector`:
  * a cache for verified approvals and a cache for unverified approvals 
  * while the collector has `processable = false` all approvals go in the cache for unverified ones
  * when the collector changes to `processable = true`, we verify the approvals (potentially concurrently) and then move them over into the verified cache as well as account for them in the `ApprovalCollector`s 
* Option (ii): Alternatively, we could abstract `AssignmentCollector` by an interface and have 3 different implementations, one for each state:
   * In state 1., the implementation only caches the approvals without verification 
   * In state 2. the implementation verifies all approvals 
   * In state 3. all operations on the `AssignmentCollector` are NoOps
* Option (iii): we change the boolean state to a enum (well Go's abomination of enums)

**I am happy to leave this as tech-debt task for a later PR.**